tagless-final/src/main/java/com/github/jorgecastillo/kotlinandroid/tagless/algebras/ui/Presentation.kt

```diff
@@ -0,0 +1,78 @@
+package com.github.jorgecastillo.kotlinandroid.tagless.algebras.ui
+
+import android.content.Context
+import arrow.HK
+import arrow.TC
+import arrow.typeclass
+import arrow.typeclasses.MonadError
+import arrow.typeclasses.binding
+import com.github.jorgecastillo.kotlinandroid.tagless.algebras.services.HeroesService
+import com.github.jorgecastillo.kotlinandroid.tagless.algebras.services.model.CharacterError
+import com.github.jorgecastillo.kotlinandroid.tagless.algebras.ui.model.SuperHeroViewModel
+import com.karumi.marvelapiclient.model.CharacterDto
+import com.karumi.marvelapiclient.model.MarvelImage
+
+interface SuperHeroesView {
+
+    fun showNotFoundError(): Unit
+
+    fun showGenericError(): Unit
+
+    fun showAuthenticationError(): Unit
+}
+
+interface SuperHeroesListView : SuperHeroesView {
+
+    fun drawHeroes(heroes: List<SuperHeroViewModel>): Unit
+
+}
+
+interface SuperHeroDetailView : SuperHeroesView {
+
+    fun drawHero(hero: SuperHeroViewModel)
+
+}
+
+@typeclass
+interface Presentation<F> : TC {
+
+    fun navigation(): Navigation<F>
+
+    fun heroesService(): HeroesService<F>
+
+    fun ME(): MonadError<F, Throwable>
+
+    fun onHeroListItemClick(ctx: Context, heroId: String): HK<F, Unit> =
+            navigation().goToHeroDetailsPage(ctx, heroId)
+
+    fun displayErrors(view: SuperHeroesView, t: Throwable): HK<F, Unit> =
+            ME().pure(when (CharacterError.fromThrowable(t)) {
+                is CharacterError.NotFoundError -> view.showNotFoundError()
+                is CharacterError.UnknownServerError -> view.showGenericError()
+                is CharacterError.AuthenticationError -> view.showAuthenticationError()
+            })
+
+    fun drawSuperHeroes(view: SuperHeroesListView): HK<F, Unit> =
+            ME().binding {
+                val result = ME().handleError(heroesService().getHeroes(), { displayErrors(view, it); emptyList() }).bind()
+                ME().pure(view.drawHeroes(result.map {
+                    SuperHeroViewModel(
+                            it.id,
+                            it.name,
+                            it.thumbnail.getImageUrl(MarvelImage.Size.PORTRAIT_UNCANNY),
+                            it.description)
+                }))
+            }
+
+
+    fun drawSuperHeroDetails(heroId: String, view: SuperHeroDetailView): HK<F, Unit> =
+            ME().binding {
+                val result = ME().handleError(heroesService().getHeroDetails(heroId), { displayErrors(view, it); CharacterDto() }).bind()
```

Comment:
The side-effect `displayErrors` has to happen on the main thread, else it'll throw an exception.

Reply:
It happens in that thread. The underlying functions implemented in the views are wrapped in `runOnUiThread`

Reply:
Yeah all the UI method are forced to run back on UI thread now, even if that style feels a bit Rusty. I would wait for @pakoito effects PR that also fixes the bindings and bindIn() to work properly with CoroutineContexts like the UI one, to achieve the same.